File: engine/src/correction_lut.cpp
#include "correction_lut.h"

#include <cmath>
#include <sstream>
#include <string>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

CorrectionLUT::CorrectionLUT() = default;

int CorrectionLUT::angleIndex(float x, float y) {
    if (x == 0.0f && y == 0.0f) return 0;
    float deg = std::atan2(y, x) * (180.0f / static_cast<float>(M_PI));
    int idx = static_cast<int>(std::round(deg)) % BINS;
    if (idx < 0) idx += BINS;
    return idx;
}

void CorrectionLUT::apply(float rx, float ry, float& cx, float& cy) const {
    if (strength_ <= 0.0f) {
        cx = rx;
        cy = ry;
        return;
    }

    float r = std::sqrt(rx * rx + ry * ry);
    if (r < 1e-6f) {
        cx = rx;
        cy = ry;
        return;
    }

    const LutEntry& e = entries_[angleIndex(rx, ry)];
    float theta = std::atan2(ry, rx);
    float corrAngleRad = e.angle_offset * static_cast<float>(M_PI) / 180.0f;
    float corrTheta = theta + corrAngleRad;
    float corrR = r * e.radius_scale;

    float bx = corrR * std::cos(corrTheta);
    float by = corrR * std::sin(corrTheta);
    bx -= e.x_cross * ry;
    by -= e.y_cross * rx;

    cx = rx + (bx - rx) * strength_;
    cy = ry + (by - ry) * strength_;

    auto clamp = [](float v) {
        return v < -1.0f ? -1.0f : (v > 1.0f ? 1.0f : v);
    };
    cx = clamp(cx);
    cy = clamp(cy);
}

void CorrectionLUT::setEntry(int deg, const LutEntry& e) {
    if (deg < 0 || deg >= BINS) return;
    entries_[deg] = e;
}

const LutEntry& CorrectionLUT::getEntry(int deg) const {
    return entries_[deg >= 0 && deg < BINS ? deg : 0];
}

void CorrectionLUT::smooth(int window) {
    auto copy = entries_;
    int half = window / 2;
    for (int i = 0; i < BINS; ++i) {
        float rs = 0.0f, ao = 0.0f, xc = 0.0f, yc = 0.0f, w = 0.0f;
        for (int d = -half; d <= half; ++d) {
            int j = (i + d + BINS) % BINS;
            float weight = 1.0f - std::abs(d) / static_cast<float>(half + 1);
            rs += copy[j].radius_scale * weight;
            ao += copy[j].angle_offset * weight;
            xc += copy[j].x_cross * weight;
            yc += copy[j].y_cross * weight;
            w += weight;
        }
        entries_[i] = {rs / w, ao / w, xc / w, yc / w};
    }
}

std::string CorrectionLUT::toJson() const {
    std::ostringstream ss;
    ss << "{\"strength\":" << strength_ << ",\"entries\":[";
    for (int i = 0; i < BINS; ++i) {
        const auto& e = entries_[i];
        ss << "{\"rs\":" << e.radius_scale
           << ",\"ao\":" << e.angle_offset
           << ",\"xc\":" << e.x_cross
           << ",\"yc\":" << e.y_cross << "}";
        if (i + 1 < BINS) ss << ",";
    }
    ss << "]}";
    return ss.str();
}

bool CorrectionLUT::fromJson(const std::string& json) {
    auto readFloat = [](const std::string& text, const std::string& key, float& out) {
        auto pos = text.find('"' + key + '"');
        if (pos == std::string::npos) return false;
        pos = text.find(':', pos);
        if (pos == std::string::npos) return false;
        try {
            out = std::stof(text.substr(pos + 1));
            return true;
        } catch (...) {
            return false;
        }
    };

    auto pos = json.find("\"strength\"");
    if (pos != std::string::npos) {
        pos = json.find(':', pos);
        if (pos != std::string::npos) {
            try {
                strength_ = std::stof(json.substr(pos + 1));
            } catch (...) {
            }
        }
    }

    auto arrPos = json.find("\"entries\"");
    if (arrPos == std::string::npos) return true;
    arrPos = json.find('[', arrPos);
    if (arrPos == std::string::npos) return false;

    int binIdx = 0;
    size_t cursor = arrPos + 1;
    while (binIdx < BINS && cursor < json.size()) {
        auto objStart = json.find('{', cursor);
        if (objStart == std::string::npos) break;

        int depth = 0;
        size_t objEnd = std::string::npos;
        for (size_t i = objStart; i < json.size(); ++i) {
            if (json[i] == '{') ++depth;
            else if (json[i] == '}' && --depth == 0) {
                objEnd = i;
                break;
            }
        }
        if (objEnd == std::string::npos) break;

        std::string obj = json.substr(objStart, objEnd - objStart + 1);
        LutEntry entry;
        readFloat(obj, "rs", entry.radius_scale);
        readFloat(obj, "ao", entry.angle_offset);
        readFloat(obj, "xc", entry.x_cross);
        readFloat(obj, "yc", entry.y_cross);
        setEntry(binIdx++, entry);
        cursor = objEnd + 1;
    }

    return binIdx > 0;
}

File: gui/Models/CorrectionLUT.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace VRStickScope.Models;

public class LutEntry
{
    [JsonPropertyName("rs")]
    public float RadiusScale { get; set; } = 1f;
    [JsonPropertyName("ao")]
    public float AngleOffset { get; set; } = 0f;
    [JsonPropertyName("xc")]
    public float XCross      { get; set; } = 0f;
    [JsonPropertyName("yc")]
    public float YCross      { get; set; } = 0f;
}

public class CorrectionLUT
{
    public const int Bins = 360;
    [JsonPropertyName("entries")]
    public LutEntry[] Entries { get; set; } = CreateDefaultEntries();
    [JsonPropertyName("strength")]
    public float Strength { get; set; } = 1f;

    private static LutEntry[] CreateDefaultEntries()
    {
        var entries = new LutEntry[Bins];
        for (int i = 0; i < Bins; i++) entries[i] = new LutEntry();
        return entries;
    }

    public void EnsureComplete()
    {
        if (Entries.Length != Bins)
        {
            var resized = CreateDefaultEntries();
            Array.Copy(Entries, resized, Math.Min(Entries.Length, Bins));
            Entries = resized;
        }

        for (int i = 0; i < Entries.Length; i++)
        {
            Entries[i] ??= new LutEntry();
        }
    }

    public static int AngleIndex(float x, float y)
    {
        if (x == 0 && y == 0) return 0;
        float deg = MathF.Atan2(y, x) * (180f / MathF.PI);
        int idx = (int)MathF.Round(deg) % 360;
        if (idx < 0) idx += 360;
        return idx;
    }

    public (float cx, float cy) Apply(float rx, float ry)
    {
        EnsureComplete();
        float r = MathF.Sqrt(rx * rx + ry * ry);
        if (r < 1e-6f) return (rx, ry);

        int idx = AngleIndex(rx, ry);
        var e = Entries[idx];

        float theta = MathF.Atan2(ry, rx);
        float corrTheta = theta + e.AngleOffset * (MathF.PI / 180f);
        float corrR = r * e.RadiusScale;

        float bx = corrR * MathF.Cos(corrTheta) - e.XCross * ry;
        float by = corrR * MathF.Sin(corrTheta) - e.YCross * rx;

        float cx = Math.Clamp(rx + (bx - rx) * Strength, -1f, 1f);
        float cy = Math.Clamp(ry + (by - ry) * Strength, -1f, 1f);
        return (cx, cy);
    }

    // 円回し診断: 収集したサンプルからLUTを構築
    public void BuildFromCircleSweep(System.Collections.Generic.List<(float rx, float ry)> samples)
    {
        // 各角度ビンへのサンプル集計
        var binSamples = new System.Collections.Generic.List<(float rx, float ry)>[Bins];
        for (int i = 0; i < Bins; i++) binSamples[i] = new();

        foreach (var (rx, ry) in samples)
        {
            float r = MathF.Sqrt(rx * rx + ry * ry);
            if (r < 0.3f) continue; // スティック中央付近は無視
            int idx = AngleIndex(rx, ry);
            binSamples[idx].Add((rx, ry));
        }

        // 各ビンの平均半径と角度誤差を計算
        float globalAvgR = 0;
        int countR = 0;
        for (int i = 0; i < Bins; i++)
        {
            if (binSamples[i].Count == 0) continue;
            foreach (var (rx, ry) in binSamples[i])
            {
                globalAvgR += MathF.Sqrt(rx * rx + ry * ry);
                countR++;
            }
        }
        if (countR > 0) globalAvgR /= countR;

        for (int i = 0; i < Bins; i++)
        {
            if (binSamples[i].Count == 0)
            {
                Entries[i] = new LutEntry();
                continue;
            }

            float avgR = 0;
            float sumSinErr = 0, sumCosErr = 0;
            float expectedAngle = i * MathF.PI / 180f;

            foreach (var (rx, ry) in binSamples[i])
            {
                float r = MathF.Sqrt(rx * rx + ry * ry);
                float actualAngle = MathF.Atan2(ry, rx);
                float angErr = expectedAngle - actualAngle;
                avgR += r;
                sumSinErr += MathF.Sin(angErr);
                sumCosErr += MathF.Cos(angErr);
            }
            avgR /= binSamples[i].Count;
            float angleErr = MathF.Atan2(sumSinErr / binSamples[i].Count, sumCosErr / binSamples[i].Count);

            Entries[i] = new LutEntry
            {
                RadiusScale = globalAvgR > 0 ? globalAvgR / avgR : 1f,
                AngleOffset = angleErr * 180f / MathF.PI,
                XCross = 0f,
                YCross = 0f
            };
        }

        // スムージング
        Smooth(5);
    }

    // 自動セクター修復: 収集したサンプルから局所的な縮退を修正するLUTを構築
    public void BuildAutomaticSectorRepair(System.Collections.Generic.List<(float rx, float ry)> samples)
    {
        var binSamples = new System.Collections.Generic.List<(float rx, float ry)>[Bins];
        for (int i = 0; i < Bins; i++) binSamples[i] = new();

        foreach (var (rx, ry) in samples)
        {
            float r = MathF.Sqrt(rx * rx + ry * ry);
            if (r < 0.2f) continue;
            int idx = AngleIndex(rx, ry);
            binSamples[idx].Add((rx, ry));
        }

        float[] binAvgR = new float[Bins];
        float[] binMaxR = new float[Bins];
        float globalMaxR = 0f;

        for (int i = 0; i < Bins; i++)
        {
            if (binSamples[i].Count == 0) continue;
            float sum = 0, max = 0;
            foreach (var (rx, ry) in binSamples[i])
            {
                float r = MathF.Sqrt(rx * rx + ry * ry);
                sum += r;
                if (r > max) max = r;
            }
            binAvgR[i] = sum / binSamples[i].Count;
            binMaxR[i] = max;
            if (max > globalMaxR) globalMaxR = max;
        }

        float[] targetR = new float[Bins];
        for (int i = 0; i < Bins; i++)
        {
            float localMax = 0;
            for (int d = -15; d <= 15; d++)
            {
                int j = (i + d + Bins) % Bins;
                if (binMaxR[j] > localMax) localMax = binMaxR[j];
            }
            targetR[i] = localMax > 0 ? localMax : (globalMaxR > 0 ? globalMaxR : 1f);
        }

        for (int i = 0; i < Bins; i++)
        {
            Entries[i] = new LutEntry();
            if (binSamples[i].Count > 0 && targetR[i] > 0 && binMaxR[i] > 0)
            {
                float scale = targetR[i] / binMaxR[i];
                scale = Math.Clamp(scale, 1f, 1.5f);
                Entries[i].RadiusScale = scale;
            }
            else
            {
                Entries[i].RadiusScale = 1f;
            }
        }

        Smooth(15);
    }

    // 補正後入力も使う局所補正: 既存補正を入れても弱い方向を追加で強める
    public void BuildAutomaticSectorRepairFromCorrected(System.Collections.Generic.List<(float rx, float ry, float cx, float cy)> samples)
    {
        EnsureComplete();
        var rawSamples = new System.Collections.Generic.List<(float rx, float ry)>();
        var binCorrectedMaxR = new float[Bins];
        var binRawMaxR = new float[Bins];
        float globalCorrectedMaxR = 0f;

        foreach (var (rx, ry, cx, cy) in samples)
        {
            float rawR = MathF.Sqrt(rx * rx + ry * ry);
            float correctedR = MathF.Sqrt(cx * cx + cy * cy);
            if (rawR < 0.2f) continue;

            rawSamples.Add((rx, ry));
            int idx = AngleIndex(rx, ry);
            if (rawR > binRawMaxR[idx]) binRawMaxR[idx] = rawR;
            if (correctedR > binCorrectedMaxR[idx]) binCorrectedMaxR[idx] = correctedR;
            if (correctedR > globalCorrectedMaxR) globalCorrectedMaxR = correctedR;
        }

        if (rawSamples.Count == 0)
        {
            return;
        }

        float[] targetR = new float[Bins];
        for (int i = 0; i < Bins; i++)
        {
            float localMax = 0;
            for (int d = -15; d <= 15; d++)
            {
                int j = (i + d + Bins) % Bins;
                if (binCorrectedMaxR[j] > localMax) localMax = binCorrectedMaxR[j];
            }
            targetR[i] = localMax > 0 ? localMax : (globalCorrectedMaxR > 0 ? globalCorrectedMaxR : 1f);
        }

        for (int i = 0; i < Bins; i++)
        {
            if (targetR[i] > 0 && binCorrectedMaxR[i] > 0 && binRawMaxR[i] > 0)
            {
                float correctedScale = targetR[i] / binCorrectedMaxR[i];
                float rawScale = targetR[i] / binRawMaxR[i];
                float scale = MathF.Max(correctedScale, rawScale);
                Entries[i].RadiusScale = Math.Clamp(MathF.Max(Entries[i].RadiusScale, scale), 1f, 1.8f);
            }
        }

        SmoothRadiusScale(15);
    }

    private void Smooth(int window)
    {
        var copy = (LutEntry[])Entries.Clone();
        int half = window / 2;
        for (int i = 0; i < Bins; i++)
        {
            float rs = 0, ao = 0, xc = 0, yc = 0, w = 0;
            for (int d = -half; d <= half; d++)
            {
                int j = ((i + d) % Bins + Bins) % Bins;
                float weight = 1f - MathF.Abs(d) / (half + 1f);
                rs += copy[j].RadiusScale * weight;
                ao += copy[j].AngleOffset * weight;
                xc += copy[j].XCross * weight;
                yc += copy[j].YCross * weight;
                w  += weight;
            }
            Entries[i] = new LutEntry { RadiusScale = rs/w, AngleOffset = ao/w, XCross = xc/w, YCross = yc/w };
        }
    }

    private void SmoothRadiusScale(int window)
    {
        var copy = (LutEntry[])Entries.Clone();
        int half = window / 2;
        for (int i = 0; i < Bins; i++)
        {
            float rs = 0, w = 0;
            for (int d = -half; d <= half; d++)
            {
                int j = ((i + d) % Bins + Bins) % Bins;
                float weight = 1f - MathF.Abs(d) / (half + 1f);
                rs += copy[j].RadiusScale * weight;
                w += weight;
            }
            Entries[i].RadiusScale = Math.Clamp(rs / w, 1f, 1.8f);
        }
    }
}
